L1T: Add jet quality and eg/tau isolation data vs. emu comparisons and ignore object index

bundocka opened the pull request:
#### PR description:

- Adds jet quality and eg/tau isolation to the CaloL2 data vs emulator plots in the online DQM
- Bug fix to ignore the ordering of CaloL2 objects for data vs emulator comparisons

#### PR validation:

- Ran locally on lxplus with recent ExpressPhysics data and checked histogram outputs are as expected
- Ran unit tests with `cmsRun DQM/Integration/python/clients/l1tstage2emulator_dqm_sourceclient-live_cfg.py unitTest=True`

bundocka committed:
L1T: Add jet quality and isolation data emu comparison, and fix object sorting bug

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40511/33710

- This PR adds an extra 16KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40511/33710/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40511/33710/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

bundocka committed:
L1T: code-checks

cmsbuild commented:
A new Pull Request was created by  @bundocka  for master.

It involves the following packages:

- DQM/L1TMonitor (**dqm**)
- DQM/L1TMonitorClient (**dqm**)


@emanueleusai, @ahmad3213, @cmsbuild, @syuvivida, @pmandrik, @micsucmed, @rvenditti can you please review it and eventually sign? Thanks.
@missirol this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1a37b8/29980/summary.html
**COMMIT**: 7ea4280b8399e10ebdbac77a672fa627a268d16f
**CMSSW**: CMSSW_13_0_X_2023-01-13-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40511/29980/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555538
 - DQMHistoTests: Total failures: 46
 - DQMHistoTests: Total nulls: 102
 - DQMHistoTests: Total successes: 3555368
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 16.999999999999996 KiB( 48 files compared)
 - DQMHistoSizes: changed ( 10024.0,... ): 0.850 KiB        L1TEMU/L1TStage2CaloLayer2
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-13-1100+1a37b8/55001/triggerResults): no differences found

emanueleusai commented on DQM/L1TMonitorClient/src/L1TStage2CaloLayer2DEClientSummary.cc:
```diff
@@ -152,7 +155,7 @@ void L1TStage2CaloLayer2DEClientSummary::processHistograms(DQMStore::IGetter &ig
     // double totalJets = 0, goodJets = 0, jetPosOff = 0, jetEtOff = 0;
```
Can remove these comments?

emanueleusai commented on DQM/L1TMonitorClient/src/L1TStage2CaloLayer2DEClientSummary.cc:
```diff
@@ -152,7 +155,7 @@ void L1TStage2CaloLayer2DEClientSummary::processHistograms(DQMStore::IGetter &ig
-    double goodRatio = 0, posOffRatio = 0, etOffRatio = 0;
+    double goodRatio = 0, posOffRatio = 0, etOffRatio = 0, qualOffRatio = 0;
 
     // hist = jetSummary_->getTH1F();
     // newHist = jetSummary->getTH1F();
```
can remove these comments?

emanueleusai commented on DQM/L1TMonitorClient/src/L1TStage2CaloLayer2DEClientSummary.cc:
```diff
@@ -179,7 +185,7 @@ void L1TStage2CaloLayer2DEClientSummary::processHistograms(DQMStore::IGetter &ig
-           isoEgEtOffRatio = 0;
+           isoEgEtOffRatio = 0, egIsoOffRatio = 0;
 
     // hist = egSummary_->getTH1F();
     // newHist = egSummary->getTH1F();
```
Can remove these comments?

emanueleusai commented on DQM/L1TMonitorClient/src/L1TStage2CaloLayer2DEClientSummary.cc:
```diff
@@ -221,7 +234,7 @@ void L1TStage2CaloLayer2DEClientSummary::processHistograms(DQMStore::IGetter &ig
-           isoTauEtOffRatio = 0;
+           isoTauEtOffRatio = 0, tauIsoOffRatio = 0;
 
     // hist = tauSummary_->getTH1F();
     // newHist = tauSummary->getTH1F();
```
Can remove these comments?

emanueleusai commented:
@bundocka there are some comments (not introduced by you) that can be deleted. It's up to you if you want to implement these now or leave them to a future PR. Other than that the PR is good to go for me (DQM plot differences exactly where we expect them)

bundocka committed:
Remove comments

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40511/33756

- This PR adds an extra 28KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40511/33756/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40511/33756/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

bundocka committed:
Fix whitespace

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40511/33758

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #40511 was updated. @emanueleusai, @ahmad3213, @cmsbuild, @syuvivida, @pmandrik, @micsucmed, @rvenditti can you please check and sign again.

emanueleusai commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1a37b8/30050/summary.html
**COMMIT**: 6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb
**CMSSW**: CMSSW_13_0_X_2023-01-17-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40511/30050/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555538
 - DQMHistoTests: Total failures: 46
 - DQMHistoTests: Total nulls: 102
 - DQMHistoTests: Total successes: 3555368
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 16.999999999999996 KiB( 48 files compared)
 - DQMHistoSizes: changed ( 10024.0,... ): 0.850 KiB        L1TEMU/L1TStage2CaloLayer2
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-17-2300+1a37b8/55072/triggerResults): no differences found

emanueleusai commented:
I was just waiting for the tests to finish after the author removed the comments.

emanueleusai commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
With the small stat visible in the bot tests for the HI wf 140.56 the "problematic" histos are now empty, which is a effect of your fix to ignore the ordering. Do you see the same (i.e. no more unmatched jets) with the supposedly larger stat that you used for your tests? Please confirm that this is expected.

Why is that bad ordering present only for jets and egamma, and not for the other objects (tau, ...) dealt with by this code? I suppose there is some bug uphill for jets and egamma, which should get better fixed rather than patching the code downhill to cope with it.

After having verified that there is this ordering issue, I now wonder if it still makes sense to fill the "Emul" plots with the object pointed by what has been verified to be the wrong index, see for example [L481-486](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L481-L486) of L1TdeStage2CaloLayer2.cc. Since you have verified that Et, Pos, Qual must be the same, you should then fill the Emul quantities with the ones extracted from `dataIt`: am I correct? The same for the corresponding Egamma plots in [L660-687](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L660-L687).

Moreover, this is a minor point but if for any of the reasons above you have to add another commit to this PR, you could profit and simplify the code in order to make it better readable and more easily maintainable. For example the whole [L429-451](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L429-L451) could be more simply rewritten as
```C++
      // object pt mismatch
      bool etGood = dataIt->hwPt() == emulIt->hwPt();
      // object position mismatch
      bool posGood = dataIt->hwPhi() == emulIt->hwPhi() && dataIt->hwEta() == emulIt->hwEta();
      // object quality mismatch
      bool qualGood = dataIt->hwQual() == emulIt->hwQual();
```
And the same for L609-632, [L808-851](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L808-L851)

perrotta commented:
hold
- @bundocka please take care of https://github.com/cms-sw/cmssw/pull/40511#issuecomment-1399212495

cmsbuild commented:
Pull request has been put on hold by  @perrotta
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

bundocka committed:
L1T: Tidy code, fix mismatch histos for emu objects

bundocka committed:
L1T: code format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40511/34112

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #40511 was updated. @emanueleusai, @cmsbuild, @syuvivida, @rvenditti, @micsucmed, @pmandrik can you please check and sign again.

bundocka commented:
> With the small stat visible in the bot tests for the HI wf 140.56 the "problematic" histos are now empty, which is a effect of your fix to ignore the ordering. Do you see the same (i.e. no more unmatched jets) with the supposedly larger stat that you used for your tests? Please confirm that this is expected.

Running over 36k events, there is perfect agreement for jets, however there are still a few unmatched EGs and ~30 unmatched taus. This is as expected. 

> Why is that bad ordering present only for jets and egamma, and not for the other objects (tau, ...) dealt with by this code? I suppose there is some bug uphill for jets and egamma, which should get better fixed rather than patching the code downhill to cope with it.

The bad ordering is present for jets, EG, and tau also. See the same fix to the tau matching here [L853-L866](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L853-L866). The mismatches arise when there are lots of low pt objects with the same et and eta, but with different phi, due to a very subtle difference in how the firmware and emulator order objects in phi. So occasionally, if there is more than 1 object with the same et and eta, but different phi, the firmware and emulator may pick different objects to be in the top 6 pt candidates for each eta half (we have a total of 12 candidates for each object, 6 for each of positive and negative eta). Quite some time was spent to try and implement the same logic in the emulator as the firmware, but it's very difficult, due to how the pipelined nature of the firmware constructs the objects as the calorimeter towers arrive into the FPGA logic. Rather than spend lots more time and effort, and make the emulator much more complicated, it was chosen to instead ignore the ordering of the objects in the list. 


> After having verified that there is this ordering issue, I now wonder if it still makes sense to fill the "Emul" plots with the object pointed by what has been verified to be the wrong index, see for example [L481-486](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L481-L486) of [[L1TdeStage2CaloLayer2.cc](http://l1tdestage2calolayer2.cc/)](http://l1tdestage2calolayer2.cc/). Since you have verified that Et, Pos, Qual must be the same, you should then fill the Emul quantities with the ones extracted from `dataIt`: am I correct? The same for the corresponding Egamma plots in [L660-687](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L660-L687).
> 

Thanks for spotting this. I have updated the code to take this into account. 

> Moreover, this is a minor point but if for any of the reasons above you have to add another commit to this PR, you could profit and simplify the code in order to make it better readable and more easily maintainable. For example the whole [L429-451](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L429-L451) could be more simply rewritten as
> 
> ```c++
>       // object pt mismatch
>       bool etGood = dataIt->hwPt() == emulIt->hwPt();
>       // object position mismatch
>       bool posGood = dataIt->hwPhi() == emulIt->hwPhi() && dataIt->hwEta() == emulIt->hwEta();
>       // object quality mismatch
>       bool qualGood = dataIt->hwQual() == emulIt->hwQual();
> ```
> 
> And the same for L609-632, [L808-851](https://github.com/cms-sw/cmssw/blob/6ab2a5b7b43647b3e4ead820a7dc5ae5f308aafb/DQM/L1TMonitor/src/L1TdeStage2CaloLayer2.cc#L808-L851)

Thanks, I've implemented your suggestion.

perrotta commented:
please test

bundocka committed:
L1T: Bugfix to emulated object iterator

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40511/34115

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #40511 was updated. @emanueleusai, @cmsbuild, @syuvivida, @rvenditti, @micsucmed, @pmandrik can you please check and sign again.

emanueleusai commented:
please test

emanueleusai commented:
type l1t

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1a37b8/30480/summary.html
**COMMIT**: 2d826acd63e568b09586f6deadb41d9f1b04a0c9
**CMSSW**: CMSSW_13_0_X_2023-02-07-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40511/30480/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 13 lines to the logs
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555852
 - DQMHistoTests: Total failures: 46
 - DQMHistoTests: Total nulls: 102
 - DQMHistoTests: Total successes: 3555682
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 16.999999999999996 KiB( 48 files compared)
 - DQMHistoSizes: changed ( 10024.0,... ): 0.850 KiB        L1TEMU/L1TStage2CaloLayer2
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-02-07-2300+1a37b8/55406/triggerResults): no differences found

perrotta commented:
unhold

emanueleusai commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
